File: apps/android/app/src/main/java/com/twende/app/data/repository/JourneyRepository.kt
package com.twende.app.data.repository

import com.twende.app.data.api.TwendeApi
import com.twende.app.data.model.Journey
import com.twende.app.data.model.JourneyTracking
import com.twende.app.data.model.Route
import com.twende.app.data.model.SeatAvailability
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class JourneyRepository @Inject constructor(
    private val api: TwendeApi,
) {
    suspend fun getRoutes(): Result<List<Route>> = runCatching {
        val response = api.getRoutes()
        if (response.isSuccessful) {
            response.body()?.data ?: emptyList()
        } else {
            throw Exception("Failed to load routes: ${response.code()}")
        }
    }

    suspend fun searchJourneys(
        routeId: String? = null,
        from: String? = null,
        to: String? = null,
        date: String? = null,
        status: String? = null,
        page: Int? = null,
        pageSize: Int? = null,
    ): Result<List<Journey>> = runCatching {
        val response = api.getJourneys(
            from = from,
            to = to,
            date = date,
            status = status,
            page = page,
            pageSize = pageSize,
        )
        if (response.isSuccessful) {
            val journeys = response.body()?.data ?: emptyList()
            if (routeId != null) {
                journeys.filter { it.routeId == routeId }
            } else {
                journeys
            }
        } else {
            throw Exception("Failed to search journeys: ${response.code()}")
        }
    }

    suspend fun getSeats(journeyId: String): Result<List<SeatAvailability>> = runCatching {
        val response = api.getSeats(journeyId)
        if (response.isSuccessful) {
            val body = response.body()!!
            if (body.success && body.data != null) {
                body.data
            } else {
                throw Exception(body.error?.message ?: "Failed to load seats")
            }
        } else {
            throw Exception("Failed to load seats: ${response.code()}")
        }
    }

    suspend fun getJourneyTracking(journeyId: String): Result<JourneyTracking> = runCatching {
        val response = api.getJourneyTracking(journeyId)
        if (response.isSuccessful) {
            val body = response.body()!!
            if (body.success && body.data != null) {
                body.data
            } else {
                throw Exception(body.error?.message ?: "Failed to load tracking data")
            }
        } else {
            throw Exception("Failed to load tracking data: ${response.code()}")
        }
    }

    suspend fun getPublicTracking(token: String): Result<JourneyTracking> = runCatching {
        val response = api.getPublicTracking(token)
        if (response.isSuccessful) {
            val body = response.body()!!
            if (body.success && body.data != null) {
                body.data
            } else {
                throw Exception(body.error?.message ?: "Failed to load tracking data")
            }
        } else {
            throw Exception("Failed to load tracking data: ${response.code()}")
        }
    }
}

File: apps/ios/Twende/Twende/Repositories/JourneyRepository.swift
import Foundation

final class JourneyRepository {
    static let shared = JourneyRepository()
    private let api = APIService.shared

    private init() {}

    func getRoutes() async throws -> [BusRoute] {
        let response: APIResponse<[BusRoute]> = try await api.get("routes")
        return response.data ?? []
    }

    func searchJourneys(routeId: String, date: String? = nil) async throws -> [Journey] {
        var query: [URLQueryItem] = [URLQueryItem(name: "routeId", value: routeId)]
        if let date = date {
            query.append(URLQueryItem(name: "date", value: date))
        }
        let response: PaginatedResponse<Journey> = try await api.get("journeys", queryItems: query)
        return response.data
    }

    func getSeats(journeyId: String) async throws -> [SeatAvailability] {
        let response: APIResponse<[SeatAvailability]> = try await api.get("journeys/\(journeyId)/seats")
        return response.data ?? []
    }

    func getJourneyTracking(journeyId: String) async throws -> JourneyTracking {
        let response: APIResponse<JourneyTracking> = try await api.get("tracking/journey/\(journeyId)")
        guard let data = response.data else {
            throw APIError.noData
        }
        return data
    }

    func getPublicTracking(token: String) async throws -> JourneyTracking {
        let response: APIResponse<JourneyTracking> = try await api.get("tracking/public/\(token)", authenticated: false)
        guard let data = response.data else {
            throw APIError.noData
        }
        return data
    }
}
